activesupport/test/message_encryptor_test.rb

```diff
@@ -106,8 +106,118 @@ def test_messing_with_aead_values_causes_failures
     assert_aead_not_decrypted(encryptor, [text, iv, auth_tag[0..-2]] * "--")
   end
 
-  private
+  def test_metadata_purpose
+    encryptor = ActiveSupport::MessageEncryptor.new(@secret, cipher: "aes-256-gcm")
+    encrypted_message = encryptor.encrypt_and_sign("foo", purpose: "bar")
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message, purpose: "bar")
+    assert_nil encryptor.decrypt_and_verify(encrypted_message, purpose: "A completely different purpose")
+  end
+
+  def test_metadata_expires_at
+    encryptor = ActiveSupport::MessageEncryptor.new(@secret, cipher: "aes-256-gcm")
+    encrypted_message = encryptor.encrypt_and_sign("foo", expires_at: 1.year.from_now)
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message)
+    travel 2.year
+    assert_nil encryptor.decrypt_and_verify(encrypted_message)
+  end
+
+  def test_metadata_expires_in
+    encryptor = ActiveSupport::MessageEncryptor.new(@secret, cipher: "aes-256-gcm")
+    encrypted_message = encryptor.encrypt_and_sign("foo", expires_in: 1.hour)
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message)
+    travel 0.9.hour
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message)
+    travel 0.11.hour
+    assert_nil encryptor.decrypt_and_verify(encrypted_message)
+  end
+
+  def test_metadata_using_different_cipher
+    encryptor = ActiveSupport::MessageEncryptor.new(@secret, cipher: "aes-256-cbc")
+    encrypted_message = encryptor.encrypt_and_sign("foo", purpose: "bar", expires_at: 2.year.from_now, expires_in: 1.hour)
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message, purpose: "bar")
+  end
+
+  def test_favor_expires_at_over_expires_in
+    encryptor = ActiveSupport::MessageEncryptor.new(@secret, cipher: "aes-256-gcm")
+    encrypted_message = encryptor.encrypt_and_sign("foo", expires_at: 2.year.from_now, expires_in: 1.hour)
+    travel 1.year
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message)
+    travel 2.year
+    assert_nil encryptor.decrypt_and_verify(encrypted_message)
+
+    encrypted_message = encryptor.encrypt_and_sign("foo", expires_at: 2.hour.from_now, expires_in: 1.year)
+    travel 1.hour
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message)
+
+    travel 2.hour
+    assert_nil encryptor.decrypt_and_verify(encrypted_message)
+  end
+
+  def test_setting_only_purpose_turns_expiration_off
+    encryptor = ActiveSupport::MessageEncryptor.new(@secret, cipher: "aes-256-gcm")
+    encrypted_message = encryptor.encrypt_and_sign("foo", purpose: "bar")
+
+    travel 1.hour
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message, purpose: "bar")
 
+    travel 1.year
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message, purpose: "bar")
+
+    travel 100.year
+    assert_equal "foo", encryptor.decrypt_and_verify(encrypted_message, purpose: "bar")
+  end
+
+  def test_purpose_using_json_serializer
```

Comment:
👆 for more tips find @jeremy's recent cleanup PR for the Active Support cache store tests.